src/cellflow/data/_datamanager.py

```diff
@@ -431,6 +440,356 @@ def _get_condition_data(
+                linked_group, linked_cov = list(linked_covar)
+
+                if linked_cov is None:
+                    linked_arr = np.full((1, 1), null_value)
+                    linked_arr = check_shape_fn(linked_arr)
+                    perturb_covar_emb[linked_group].append(linked_arr)
+                    continue
+
+                cov_name = condition_data[linked_cov]
+
+                if linked_group in covariate_reps:
+                    rep_key = covariate_reps[linked_group]
+                    if cov_name not in rep_dict[rep_key]:
+                        raise ValueError(f"Representation for '{cov_name}' not found in `adata.uns['{linked_group}']`.")
+                    linked_arr = np.asarray(rep_dict[rep_key][cov_name])
+                else:
+                    linked_arr = np.asarray(condition_data[linked_cov])
+
+                linked_arr = check_shape_fn(linked_arr)
+                perturb_covar_emb[linked_group].append(linked_arr)
+
+        perturb_covar_emb = {
+            k: pad_to_max_length_fn(
+                np.concatenate(v, axis=0),
+                max_combination_length,
+                null_value,
+            )
+            for k, v in perturb_covar_emb.items()
+        }
+
+        sample_covar_emb: dict[str, np.ndarray] = {}
+        for sample_cov in sample_covariates:
+            value = condition_data[sample_cov]
+            if sample_cov in covariate_reps:
+                rep_key = covariate_reps[sample_cov]
+
+                if value not in rep_dict[rep_key]:
+                    raise ValueError(f"Representation for '{value}' not found in `adata.uns['{sample_cov}']`.")
+                cov_arr = np.asarray(rep_dict[rep_key][value])
+            else:
+                cov_arr = np.asarray(value)
+
+            cov_arr = check_shape_fn(cov_arr)
+            sample_covar_emb[sample_cov] = np.tile(cov_arr, (max_combination_length, 1))
+
+        return perturb_covar_emb | sample_covar_emb
+
+    def _get_condition_data(
+        self,
+        split_cov_combs: np.ndarray | list[list[Any]],
+        adata: anndata.AnnData | None,
+        covariate_data: pd.DataFrame | None = None,
+        rep_dict: dict[str, Any] | None = None,
+        condition_id_key: str | None = None,
+    ) -> ReturnData:
+        # for prediction: adata is None, covariate_data is provided
+        # for training/validation: adata is provided and used to get cell masks, covariate_data is None
+        if adata is None and covariate_data is None:
+            raise ValueError("Either `adata` or `covariate_data` must be provided.")
+        covariate_data = covariate_data if covariate_data is not None else adata.obs  # type: ignore[union-attr]
+        if (
+            len(self._split_covariates) == 0
+            or len(self._perturbation_covariates) == 0
+            or len(self._sample_covariates) == 0
+            or not self.is_conditional
+            or adata is None
+        ):
```

Comment:
why do we do this? 
are there cases which the "new" `get_condition_data` cannot handle?

Reply:
This was the only part remaining to refactor but I don't have a quick way to implement it in my mind right away. Because I use dataframes to replicate the old code when `len(self._sample_covariates) == 0` the dataframe operations fails because one table is empty. This might take a bit more time because I am not sure how the behavior is supposed to be